### DISCUSSION TITLE
Keithley 2450 For-Loop Only Executing Once

### DISCUSSION THREAD

**Discussion by RahulVedula** [pymeasure/pymeasure#678] [Q&A] [answered]
Hi, sorry for the unnecessary post, but I've been trying to figure out how to get my Keithley2450 instrument to measure some data for an experiment. However, I keep running into the same problem, the loop that should change the currents in increments, is only running once. The exit condition is not executing either, the loop just seems to run only once. However, each time it runs, it _gives the correct number_ . I'm assuming that the issue is with my code and not the Keithley instruments so I tried simplifying it even more. For more context, in this code, I'm just measuring the resistance of a resitor. Here is the code:
```
from pymeasure.instruments.keithley import Keithley2450 #import machinery
from pymeasure.experiment import Procedure, Results, Worker #procedure class

from pymeasure.experiment import IntegerParameter, FloatParameter #Parameters, with units
import numpy as np #Numpy, dynamic arrays
from time import sleep #sleep/delay computer

class Measurements (Procedure):
   
    data_points = IntegerParameter("Data points", default=10)
    averages = IntegerParameter("Averages", default=50)
    max_current = FloatParameter("Maximum Current", units='mA', default=0.001)
    min_current = FloatParameter("Minimum Current", units='mA', default=0.0005)
   
   
    DATA_COLUMNS = ['Current (A)', 'Voltage (V)', 'Voltage Std (V)']
   
    def startup(self):
        print ("a")
        self.keithley = Keithley2450("GPIB::18")
        self.keithley.reset()
        self.keithley.apply_current()                
        self.keithley.source_current_range = 10e-3  
        self.keithley.compliance_voltage = 10        
        self.keithley.source_current = 0            
        self.keithley.enable_source()                
        #self.keithley.buffer_points(data_points)
```

    def execute(self):
        print ("a")
        currents = np.linspace(
            self.min_current,
            self.max_current,
            num=self.data_points
        )
        for i in currents:
            self.keithley.start_buffer()
            sleep(0.1)
            self.keithley.source_current = i  
            self.keithley.measure_voltage()          
            print(self.keithley.voltage)
            print(self.keithley.std_voltage)
            print(self.keithley.mean_voltage)
            self.emit('results', {
                'Current (A)': i,
                'Voltage (V)': self.keithley.mean_voltage,
                'Voltage Std (V)': self.keithley.voltage
            })
            sleep(0.01)

            if self.should_stop():
                break
           

    def shutdown(self):
        print ("a")
        self.keithley.disable_buffer()
        self.keithley.shutdown()                    
       
    if __name__ == "__main__":
    procedure = Measurements()
    procedure.data_points = 10
    procedure.averages = 50
    procedure.max_current = 0.001
    procedure.min_current = 0.0005
    data_filename = 'classCode1.csv'
    results = Results(procedure, data_filename)

    worker = Worker(results)
    worker.start()
    worker.join(timeout=3600) # wait at most 1 hr (3600 sec)`****

**Comment by bilderbuchi** [comment, marked as answer]
What is printing when you run this? Can you add a "print(i)" in the loop?

**Reply by RahulVedula** [reply]
Sorry forgot to include this earlier, but I tried this and the result was the minimum current (which is correct). It also only executed once. For example,

0.0005 (current)
10.0025 (voltage)

*Loop terminated*

**Reply by BenediktBurger** [reply]
So it does not print the standard deviation of the voltage?

I don't know exactly how those buffer in the Keithley instruments work, but I guess, that is the problem.

Reading the last voltage is possible, but std_voltage tries to read the standard deviation of the buffer. Maybe the measurement has not yet finished filling the buffer?

Do these commands work, if you use them singly in a command prompt?

**Reply by RahulVedula** [reply]
Thanks for the help. In response to your question about the command prompt—they do not work. 

Any ideas on whats making the buffer not fill?

**Reply by BenediktBurger** [reply]
I don't know the device, I just looked at the code. I'm sure the manual will tell how the device works. You can look at the code, get the command sent to the device and look up that command in the manual.
Most of the time, device manufacturers offer quite decent instructions.
